appscale/tools/agents/azure_agent.py

```diff
@@ -0,0 +1,689 @@
+                                  "use a different keyname.")
+
+    LocalState.generate_rsa_key(keyname, parameters[self.PARAM_VERBOSE])
+
+    AppScaleLogger.log("Configuring network for machine/s under "
+                       "resource group '{0}' with storage account '{1}' "
+                       "in zone '{2}'".format(resource_group, storage_account, zone))
+    # Create a resource group and an associated storage account to access resources.
+    self.create_resource_group(parameters, credentials)
+
+    resource_client = ResourceManagementClient(credentials, subscription_id)
+    resource_client.providers.register('Microsoft.Compute')
+    resource_client.providers.register('Microsoft.Network')
+
+  def describe_instances(self, parameters, pending=False):
+    """Queries Microsoft Azure to see which instances are currently
+    running, and retrieves information about their public and private IPs.
+    Args:
+      parameters: A dict containing values necessary to authenticate with the
+        underlying cloud.
+      pending: If we should show pending instances.
+    Returns:
+      A tuple of the form (public, private, id) where public is a list
+      of private IP addresses, private is a list of private IP addresses,
+      and id is a list of Azure VM names.
+    """
+    credentials = self.open_connection(parameters)
+    subscription_id = parameters[self.PARAM_SUBCR_ID]
+    resource_group = parameters[self.PARAM_RESOURCE_GROUP]
+    network_client = NetworkManagementClient(credentials, subscription_id)
+    compute_client = ComputeManagementClient(credentials, subscription_id)
+    public_ips = []
+    private_ips = []
+    instance_ids = []
+
+    public_ip_addresses = network_client.public_ip_addresses.list(resource_group)
+    for public_ip in public_ip_addresses:
+      public_ips.append(public_ip.ip_address)
+
+    network_interfaces = network_client.network_interfaces.list(resource_group)
+    for network_interface in network_interfaces:
+      for ip_config in network_interface.ip_configurations:
+        private_ips.append(ip_config.private_ip_address)
+
+    virtual_machines = compute_client.virtual_machines.list(resource_group)
+    for vm in virtual_machines:
+      instance_ids.append(vm.name)
+    return public_ips, private_ips, instance_ids
+
+  def run_instances(self, count, parameters, security_configured):
+    """ Starts 'count' instances in Microsoft Azure, and returns once they
+    have been started. Callers should create a network and attach a firewall
+    to it before using this method, or the newly created instances will not
+    have a network and firewall to attach to (and thus this method will fail).
+    Args:
+      count: An int, that specifies how many virtual machines should be started.
+      parameters: A dict, containing all the parameters necessary to
+        authenticate this user with Azure.
+      security_configured: Unused, as we assume that the network and firewall
```

Comment:
Can't we pass in security_configured=False and call the corresponding method here?

If this argument is not used by any other agent, then we should remove it IMHO.